sycl/doc/extensions/proposed/sycl_ext_oneapi_virtual_functions.asciidoc

```diff
@@ -0,0 +1,683 @@
+any other translation unit that declares the same function must also decorate
+the function with the same property (with the same argument).
+
+The programs that decorate the same function with multiple instances of the
+property with different argument are ill formed.
+|`calls_indirectly`
+|The `calls_indirectly` property indicates that a SYCL kernel function is
+performing calls through virtual member functions and declares use of one or
+more sets of virtual member functions.
+
+Optional parameter `SetIds` specifies which sets of virtual member functions are
+declared to be used by a kernel, it must be zero or more C\++ typenames. If the
+argument is omitted (zero C++ typenames specified), then a kernel is considered
+to be using the default set of virtual member functions.
+
+Calling a virtual member function, which does not belong to any of sets of
+virtual member functions declared to be used is an undefined behavior.
+
+This property should be attached to a kernel if it contains a virtual member
+function call in its call graph, even if the said function is never actually
+called. If a kernel submitted without this property calls a virtual member
+function, diagnostic should be emitted by an implementation.
+|===
+
+If a kernel is submitted with the `indirectly_callable` property that points to
+an empty set of virtual functions, an exception should be thrown by an
+implementation.
+
+Applying the `indirectly_callable` property to a SYCL Kernel function is illegal
+and an implementation should produce a diagnostic for that.
+
+Applying the `indirectly_callable` property to an arbitrary device function,
+which is not a virtual member function has no effect.
+
+NOTE: This behavior may be changed in either future version of this extension or
+in another extensions.
+
+Virtual member functions that are decorated with the `indirectly_callable`
+property are considered to be device functions, i.e. they  must obey the
+restrictions listed in section 5.4 of the core SYCL specification "Language
+restrictions for device functions". Virtual member functions that are not
+decorated with this attribute do not need to obey these restrictions, even if
+other definitions of that virtual member function in other classes in the
+inheritance hierarchy are decorated with the attribute.
+
+[source,dpcpp]
+----
+using syclext = sycl::ext::oneapi::experimental;
+
+struct set_A;
+struct set_B;
+
+class Foo {
+public:
+  // properties to functions should be applied using the macro:
+  virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
+      syclext::indirectly_callable<set_A>) void
+  foo() {}
+
+  virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
+      syclext::indirectly_callable<set_A>) void
+  bar(){};
+
+  // both declaration and definition should be annotated:
+  virtual SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(
+      syclext::indirectly_callable<set_B>) void
+  baz();
+};
+
+// both declaration and definition should be annotated:
```

Comment:
The description above for `indirectly_callable` says that you do *not* need to annotated both declaration and definition.  It should be sufficient to annotate only the first declaration.  Remember, a function definition is also a declaration.  Therefore, the definition is a "redeclaration" of the function.  We say above that this does not need to repeat the property:

> Redeclarations of the function may optionally be decorated with the same property if the property argument is the same. The effect is the same regardless of whether redeclarations are so decorated.